File: src/simulation/construction/ConstructionSites.js
/**
 * ConstructionSites — helpers for the v0.8.4 construction-in-progress system.
 *
 * Authoritative state for any in-progress build/demolish lives on the
 * tile's `tileState.construction` overlay (see plan §1.3). This module
 * mirrors that into `state.constructionSites`, an indexed array used by
 *   - WorkerAISystem.handleConstruct → quickly find/assign a builder site
 *   - ConstructionSystem.update      → tick completion checks
 *   - SceneRenderer                  → draw blueprint meshes
 *   - RoleAssignmentSystem           → size the BUILDER quota
 *
 * The mirror is recomputed via small mutators (push/splice) plus a full
 * rebuild that scans tileState (used after snapshot load).
 */
import { BALANCE } from "../../config/balance.js";
import { TILE } from "../../config/constants.js";
import { getTileState, setTileField, toIndex } from "../../world/grid/Grid.js";

/**
 * Returns the BALANCE work-time (seconds) required for the given build
 * tool, or a sensible default. Roads / light tiles complete fast, large
 * buildings take longer.
 *
 * @param {string} tool
 * @returns {number}
 */
export function getBuildWorkSec(tool) {
  const table = BALANCE.constructionWorkSec ?? {};
  const v = Number(table[tool]);
  if (Number.isFinite(v) && v > 0) return v;
  return Number(table.default ?? 4.0);
}

/**
 * Returns the BALANCE work-time (seconds) required to demolish the given
 * old tile. Ruins clear fastest; walls/gates are intermediate; built
 * structures fall back to `default`.
 *
 * @param {number} oldTile — TILE.* the demolisher is removing
 * @returns {number}
 */
export function getDemolishWorkSec(oldTile) {
  const table = BALANCE.demolishWorkSec ?? {};
  if (oldTile === TILE.RUINS) return Number(table.ruins ?? 1.5);
  if (oldTile === TILE.WALL) return Number(table.wall ?? 2.5);
  if (oldTile === TILE.GATE) return Number(table.gate ?? 2.5);
  return Number(table.default ?? 3.0);
}

/**
 * Push a new construction-site index entry. Caller is responsible for
 * having already written the authoritative overlay to
 * `tileState.get(idx).construction`. We just maintain the mirror.
 *
 * @param {object} state
 * @param {object} entry — { ix, iz, kind, tool, ... }
 */
export function pushConstructionSite(state, entry) {
  state.constructionSites ??= [];
  state.constructionSites.push(entry);
}

/**
 * Remove the construction-site index entry at (ix, iz), if any. The
 * authoritative overlay on tileState should already be cleared by the
 * caller (or be on its way out).
 *
 * @param {object} state
 * @param {number} ix
 * @param {number} iz
 * @returns {object|null} the removed entry, or null if none
 */
export function spliceConstructionSite(state, ix, iz) {
  const arr = Array.isArray(state.constructionSites) ? state.constructionSites : null;
  if (!arr) return null;
  for (let i = 0; i < arr.length; i += 1) {
    const s = arr[i];
    if (s && s.ix === ix && s.iz === iz) {
      arr.splice(i, 1);
      return s;
    }
  }
  return null;
}

/**
 * Find an existing site at (ix, iz). Returns null if none.
 *
 * @param {object} state
 * @param {number} ix
 * @param {number} iz
 * @returns {object|null}
 */
export function findConstructionSite(state, ix, iz) {
  const arr = Array.isArray(state.constructionSites) ? state.constructionSites : null;
  if (!arr) return null;
  for (const s of arr) {
    if (s && s.ix === ix && s.iz === iz) return s;
  }
  return null;
}

/**
 * Rebuild the entire `state.constructionSites` array from `tileState`.
 * Used after snapshot restore (the overlay survives, but the index does
 * not).
 *
 * @param {object} state
 */
export function rebuildConstructionSites(state) {
  const out = [];
  const grid = state?.grid;
  if (grid?.tileState?.forEach) {
    grid.tileState.forEach((entry, idx) => {
      const overlay = entry?.construction;
      if (!overlay) return;
      const ix = idx % grid.width;
      const iz = (idx - ix) / grid.width;
      out.push({
        ix,
        iz,
        kind: overlay.kind,
        tool: overlay.tool,
        builderId: overlay.builderId ?? null,
        workAppliedSec: Number(overlay.workAppliedSec ?? 0),
        workTotalSec: Number(overlay.workTotalSec ?? 0),
      });
    });
  }
  state.constructionSites = out;
}

/**
 * Read the canonical construction overlay for a tile (or null if none).
 * Centralised here so callers don't accidentally read a stale mirror.
 *
 * @param {object} state
 * @param {number} ix
 * @param {number} iz
 * @returns {object|null}
 */
export function getConstructionOverlay(state, ix, iz) {
  const entry = getTileState(state.grid, ix, iz);
  return entry?.construction ?? null;
}

/**
 * Find an unassigned construction site for `worker`, or return the site
 * already reserved by this builder if it still exists. Mutates: assigns
 * `builderId` on both the index entry and the authoritative overlay so
 * that subsequent ticks (and other workers) respect the reservation.
 *
 * Selection heuristic: prefer the site already reserved by this worker;
 * otherwise pick the unassigned site whose Manhattan distance from the
 * worker is smallest.
 *
 * @param {object} state
 * @param {object} worker
 * @returns {object|null} site index entry, or null
 */
export function findOrReserveBuilderSite(state, worker) {
  const arr = Array.isArray(state.constructionSites) ? state.constructionSites : null;
  if (!arr || arr.length === 0) return null;
  const wx = Number(worker.x ?? 0);
  const wz = Number(worker.z ?? 0);
  const grid = state.grid;
  // Prefer worker's existing reservation
  for (const site of arr) {
    if (!site) continue;
    if (site.builderId && site.builderId === worker.id) {
      // Re-sync mirror with overlay to keep the FSM honest
      const overlay = getConstructionOverlay(state, site.ix, site.iz);
      if (overlay) {
        site.workAppliedSec = Number(overlay.workAppliedSec ?? 0);
        site.workTotalSec = Number(overlay.workTotalSec ?? 0);
      }
      return site;
    }
  }
  // Otherwise pick nearest unassigned site
  let best = null;
  let bestDist = Infinity;
  for (const site of arr) {
    if (!site) continue;
    if (site.builderId && site.builderId !== worker.id) continue;
    if (!grid) continue;
    // Approximate distance via tileToWorld math: tile centre offset from
    // worker — Manhattan in tile units is fine for selection.
    const tileCenterX = (site.ix - grid.width / 2 + 0.5) * grid.tileSize;
    const tileCenterZ = (site.iz - grid.height / 2 + 0.5) * grid.tileSize;
    const d = Math.abs(tileCenterX - wx) + Math.abs(tileCenterZ - wz);
    if (d < bestDist) {
      bestDist = d;
      best = site;
    }
  }
  if (!best) return null;
  best.builderId = worker.id;
  // Reflect on the overlay so the next system tick (or another worker)
  // sees the reservation.
  const overlay = getConstructionOverlay(state, best.ix, best.iz);
  if (overlay) overlay.builderId = worker.id;
  return best;
}

/**
 * Release any builder reservation held by this worker (e.g. on death,
 * role change, or completion).
 *
 * @param {object} state
 * @param {object} worker
 */
export function releaseBuilderSite(state, worker) {
  const arr = Array.isArray(state.constructionSites) ? state.constructionSites : null;
  if (!arr) return;
  for (const site of arr) {
    if (!site) continue;
    if (site.builderId === worker.id) {
      site.builderId = null;
      const overlay = getConstructionOverlay(state, site.ix, site.iz);
      if (overlay) overlay.builderId = null;
    }
  }
}

/**
 * Increment workAppliedSec on the overlay AND mirror entry by `dt`. We
 * write to the overlay via direct property mutation (the tileState entry
 * is a plain object; setTileField is for typed fields tracked by the
 * change-version machinery, but the overlay ticks every frame and we
 * don't want to inflate `tileStateVersion` on each accumulation).
 *
 * @param {object} state
 * @param {number} ix
 * @param {number} iz
 * @param {number} dt — seconds to add
 * @returns {object|null} updated overlay, or null if missing
 */
export function applyConstructionWork(state, ix, iz, dt) {
  const overlay = getConstructionOverlay(state, ix, iz);
  if (!overlay) return null;
  overlay.workAppliedSec = Math.max(0, Number(overlay.workAppliedSec ?? 0)) + Math.max(0, Number(dt) || 0);
  // Mirror the per-frame progress into the index entry so renderer reads
  // are cheap (no tileState lookup per blueprint).
  const arr = Array.isArray(state.constructionSites) ? state.constructionSites : null;
  if (arr) {
    for (const site of arr) {
      if (site && site.ix === ix && site.iz === iz) {
        site.workAppliedSec = overlay.workAppliedSec;
        site.workTotalSec = Number(overlay.workTotalSec ?? site.workTotalSec ?? 0);
        break;
      }
    }
  }
  return overlay;
}

/**
 * Write the construction overlay onto a tile's tileState. Resources have
 * already been spent by the caller (BuildSystem.placeToolAt); this just
 * persists the metadata + bumps the tileState change version.
 *
 * @param {object} state
 * @param {number} ix
 * @param {number} iz
 * @param {object} overlay
 */
export function setConstructionOverlay(state, ix, iz, overlay) {
  setTileField(state.grid, ix, iz, "construction", overlay);
}

/**
 * Clear the overlay on a tile (called on completion or cancel).
 *
 * @param {object} state
 * @param {number} ix
 * @param {number} iz
 */
export function clearConstructionOverlay(state, ix, iz) {
  // Use setTileField with `null` so the change-version still bumps and
  // any downstream cache (renderer signature) re-evaluates.
  setTileField(state.grid, ix, iz, "construction", null);
  const grid = state?.grid;
  if (!grid?.tileState) return;
  const idx = toIndex(ix, iz, grid.width);
  const entry = grid.tileState.get(idx);
  if (entry && "construction" in entry) {
    delete entry.construction;
  }
}

File: project_utopia/simulation/construction/construction_sites.py
"""Construction-site index helpers (port of ``ConstructionSites.js``).

Authoritative state for any in-progress build/demolish lives on the
tile's overlay (stored in :attr:`Grid.tile_state`). This module mirrors
that into ``state["constructionSites"]``, an indexed list used by the
build / construction systems to assign builders and tick completion.
"""

from __future__ import annotations

from typing import Any

from project_utopia.world.grid import TILE, Grid

__all__ = [
    "get_build_work_sec",
    "get_demolish_work_sec",
    "push_construction_site",
    "splice_construction_site",
    "find_construction_site",
    "rebuild_construction_sites",
    "get_construction_overlay",
    "set_construction_overlay",
    "clear_construction_overlay",
    "apply_construction_work",
    "find_or_reserve_builder_site",
    "release_builder_site",
]


_CONSTRUCTION_WORK_SEC: dict[str, float] = {
    "road": 1.5,
    "farm": 4.0,
    "lumber": 4.0,
    "warehouse": 6.0,
    "wall": 5.0,
    "quarry": 5.0,
    "bridge": 5.0,
    "default": 4.0,
}

_DEMOLISH_WORK_SEC: dict[str, float] = {
    "ruins": 1.5,
    "wall": 2.5,
    "default": 3.0,
}


def get_build_work_sec(tool: str) -> float:
    v = _CONSTRUCTION_WORK_SEC.get(str(tool))
    if v is None or v <= 0:
        return _CONSTRUCTION_WORK_SEC["default"]
    return float(v)


def get_demolish_work_sec(old_tile: int) -> float:
    if int(old_tile) == TILE["RUINS"]:
        return float(_DEMOLISH_WORK_SEC["ruins"])
    if int(old_tile) == TILE["WALL"]:
        return float(_DEMOLISH_WORK_SEC["wall"])
    return float(_DEMOLISH_WORK_SEC["default"])


def push_construction_site(state: dict[str, Any], entry: dict[str, Any]) -> None:
    state.setdefault("constructionSites", []).append(entry)


def splice_construction_site(
    state: dict[str, Any], ix: int, iz: int
) -> dict[str, Any] | None:
    arr = state.get("constructionSites")
    if not isinstance(arr, list):
        return None
    for i, s in enumerate(arr):
        if isinstance(s, dict) and int(s.get("ix", -1)) == ix and int(s.get("iz", -1)) == iz:
            return arr.pop(i)
    return None


def find_construction_site(
    state: dict[str, Any], ix: int, iz: int
) -> dict[str, Any] | None:
    for s in state.get("constructionSites") or []:
        if isinstance(s, dict) and int(s.get("ix", -1)) == ix and int(s.get("iz", -1)) == iz:
            return s
    return None


def get_construction_overlay(
    state: dict[str, Any], ix: int, iz: int
) -> dict[str, Any] | None:
    grid = state.get("grid")
    if not isinstance(grid, Grid):
        return None
    if not hasattr(grid, "tile_state"):
        return None
    idx = iz * grid.width + ix
    entry = grid.tile_state.get(idx)
    if not isinstance(entry, dict):
        return None
    overlay = entry.get("construction")
    return overlay if isinstance(overlay, dict) else None


def set_construction_overlay(
    state: dict[str, Any], ix: int, iz: int, overlay: dict[str, Any]
) -> None:
    grid = state.get("grid")
    if not isinstance(grid, Grid):
        return
    if not hasattr(grid, "tile_state"):
        return
    idx = iz * grid.width + ix
    entry = grid.tile_state.get(idx)
    if not isinstance(entry, dict):
        entry = {}
        grid.tile_state[idx] = entry
    entry["construction"] = overlay


def clear_construction_overlay(
    state: dict[str, Any], ix: int, iz: int
) -> None:
    grid = state.get("grid")
    if not isinstance(grid, Grid):
        return
    if not hasattr(grid, "tile_state"):
        return
    idx = iz * grid.width + ix
    entry = grid.tile_state.get(idx)
    if isinstance(entry, dict) and "construction" in entry:
        del entry["construction"]


def rebuild_construction_sites(state: dict[str, Any]) -> None:
    """Rebuild ``state["constructionSites"]`` from authoritative overlays."""
    grid = state.get("grid")
    out: list[dict[str, Any]] = []
    if isinstance(grid, Grid) and hasattr(grid, "tile_state"):
        for idx in sorted(grid.tile_state.keys()):
            entry = grid.tile_state[idx]
            if not isinstance(entry, dict):
                continue
            overlay = entry.get("construction")
            if not isinstance(overlay, dict):
                continue
            ix = idx % grid.width
            iz = idx // grid.width
            out.append(
                {
                    "ix": ix,
                    "iz": iz,
                    "kind": overlay.get("kind"),
                    "tool": overlay.get("tool"),
                    "builderId": overlay.get("builderId"),
                    "workAppliedSec": float(overlay.get("workAppliedSec", 0.0)),
                    "workTotalSec": float(overlay.get("workTotalSec", 0.0)),
                }
            )
    state["constructionSites"] = out


def apply_construction_work(
    state: dict[str, Any], ix: int, iz: int, dt: float
) -> dict[str, Any] | None:
    """Increment ``workAppliedSec`` on the overlay + index entry by ``dt``."""
    overlay = get_construction_overlay(state, ix, iz)
    if overlay is None:
        return None
    overlay["workAppliedSec"] = max(0.0, float(overlay.get("workAppliedSec", 0.0))) + max(
        0.0, float(dt)
    )
    arr = state.get("constructionSites")
    if isinstance(arr, list):
        for site in arr:
            if isinstance(site, dict) and int(site.get("ix", -1)) == ix and int(site.get("iz", -1)) == iz:
                site["workAppliedSec"] = float(overlay["workAppliedSec"])
                site["workTotalSec"] = float(overlay.get("workTotalSec", site.get("workTotalSec", 0.0)))
                break
    return overlay


def find_or_reserve_builder_site(
    state: dict[str, Any], worker: dict[str, Any]
) -> dict[str, Any] | None:
    arr = state.get("constructionSites")
    if not isinstance(arr, list) or not arr:
        return None
    wid = worker.get("id")
    # Prefer existing reservation by this worker.
    for site in arr:
        if not isinstance(site, dict):
            continue
        if site.get("builderId") == wid:
            overlay = get_construction_overlay(state, int(site["ix"]), int(site["iz"]))
            if overlay is not None:
                site["workAppliedSec"] = float(overlay.get("workAppliedSec", 0.0))
                site["workTotalSec"] = float(overlay.get("workTotalSec", 0.0))
            return site
    # Otherwise pick the nearest unassigned site.
    wx = float(worker.get("x", 0.0))
    wz = float(worker.get("z", 0.0))
    best: dict[str, Any] | None = None
    best_dist = float("inf")
    for site in arr:
        if not isinstance(site, dict):
            continue
        if site.get("builderId") and site.get("builderId") != wid:
            continue
        d = abs(int(site.get("ix", 0)) - wx) + abs(int(site.get("iz", 0)) - wz)
        if d < best_dist:
            best_dist = d
            best = site
    if best is None:
        return None
    best["builderId"] = wid
    overlay = get_construction_overlay(state, int(best["ix"]), int(best["iz"]))
    if overlay is not None:
        overlay["builderId"] = wid
    return best


def release_builder_site(state: dict[str, Any], worker: dict[str, Any]) -> None:
    arr = state.get("constructionSites")
    if not isinstance(arr, list):
        return
    wid = worker.get("id")
    for site in arr:
        if not isinstance(site, dict):
            continue
        if site.get("builderId") == wid:
            site["builderId"] = None
            overlay = get_construction_overlay(state, int(site["ix"]), int(site["iz"]))
            if overlay is not None:
                overlay["builderId"] = None
